simpunch/level2.py

```diff
@@ -135,22 +138,98 @@ def generate_starfield(wcs: WCS,
                                                            input_data.meta.astropy_time,
                                                            input_data.data.shape)
 
-    starfield, stars = generate_starfield(wcs_stellar_input, input_data.data[0, :, :].shape, flux_set=2.0384547E-9,
-                                          fwhm=3, dimmest_magnitude=12, noise_mean=None, noise_std=None)
+    starfield, stars = generate_starfield(wcs_stellar_input, input_data.data.shape,
+                                          flux_set=2.0384547E-9, fwhm=3, dimmest_magnitude=12,
+                                          noise_mean=None, noise_std=None)
 
     starfield_data = np.zeros(input_data.data.shape)
-    for i in range(starfield_data.shape[0]):
-        starfield_data[i, :, :] = starfield * (np.logical_not(np.isclose(input_data.data[i, :, :], 0, atol=1E-18)))
+    starfield_data[:, :] = starfield * (np.logical_not(np.isclose(input_data.data, 0, atol=1E-18)))
 
-    input_data.data[...] = input_data.data[...] + starfield_data
+    # Converting the input data polarization to celestial basis
+    mzp_angles = ([input_cube.meta["POLAR"].value for label, input_cube in input_collection.items() if
+                   label != "alpha"]) * u.degree
+    cel_north_off = get_p_angle(time=input_collection["Z"].meta["DATE-OBS"].value)
+    new_angles = (mzp_angles + cel_north_off).value * u.degree
+    # check - or + ? confirm!
 
-    return input_data
+    valid_keys = [key for key in input_collection if key != "alpha"]
+
+    meta_a = dict(NormalizedMetadata.to_fits_header(input_collection[valid_keys[0]].meta,
+                                                    wcs=input_collection[valid_keys[0]].wcs))
+    meta_b = dict(NormalizedMetadata.to_fits_header(input_collection[valid_keys[1]].meta,
+                                                    wcs=input_collection[valid_keys[1]].wcs))
+    meta_c = dict(NormalizedMetadata.to_fits_header(input_collection[valid_keys[2]].meta,
+                                                    wcs=input_collection[valid_keys[2]].wcs))
+
+    meta_a["POLAR"] = meta_a["POLAR"] * u.degree
+    meta_b["POLAR"] = meta_b["POLAR"] * u.degree
+    meta_c["POLAR"] = meta_c["POLAR"] * u.degree
+
+    data_collection = NDCollection(
+        [(str(valid_keys[0]), NDCube(data=input_collection[valid_keys[0]].data,
+                                     meta=meta_a, wcs=input_collection[valid_keys[0]].wcs)),
+         (str(valid_keys[1]), NDCube(data=input_collection[valid_keys[1]].data,
+                                     meta=meta_b, wcs=input_collection[valid_keys[1]].wcs)),
+         (str(valid_keys[2]), NDCube(data=input_collection[valid_keys[2]].data,
+                                     meta=meta_c, wcs=input_collection[valid_keys[2]].wcs))],
+        aligned_axes="all")
+
+    input_data_cel = solpolpy.resolve(data_collection, "npol", reference_angle=0 * u.degree, out_angles=new_angles)
```

Comment:
I'm confused. Why do we have to go through npol?

Reply:
Because we are initially in mzpsol. Stars have polarization w.r.t. celestial north. So the mzpsol is converted to celestial frames and stars are added. Once done, the complete states are then transformed to mzpinstru